### PR TITLE
Add behavior test mode.

### PR DESCRIPTION
**Pull request by kevinlemonra**
./go-stun -b -s stun.stunprotocol.org:3478
Mapping Behavior:        EndpointIndependent
Filtering Behavior:         AddressAndPortDependent
Normal NAT Type:         Port Restricted cone NAT

### CODE REVIEW

**Review by ccding — COMMENTED** [review]
Thanks for the PR. Some comments are below, but at a high level, I am not quite sure the benefit of having users specify addresses while using the default port. I know `ssh` and `telnet` commands do so, but their use cases always require an address and the address could be different for each run; having a default port can let users not type a port every time. However, stun is not the same: users can use the same address/port for all runs on a particular computer and they can always copy the port while copying the address.  Can you help me understand it a little bit more?

File: stun/client.go [hunk 1]
```diff
@@ -79,9 +85,6 @@ func (c *Client) SetSoftwareName(name string) {
 // Discover contacts the STUN server and gets the response of NAT type, host
 // for UDP punching.
 func (c *Client) Discover() (NATType, *Host, error) {
-	if c.serverAddr == "" {
-		c.SetServerAddr(DefaultServerAddr)
-	}
```
[the remark this answers is not in the document]
Comment (kevinlemonra) [reply stun/client.go:84]: There is a bug need to fix.

[the remark this answers is not in the document]
Comment (ccding) [reply stun/client.go:84]: I think we should fix this before merging the PR

File: stun/const.go [hunk 2]
```diff
@@ -28,6 +29,14 @@ const (
 // NATType is the type of NAT described by int.
 type NATType int
 
+// NAT behavior type
+type NATBhvType int
+
+type NATBehavior struct {
+	MapBhvType NATBhvType // MappingBehaviorType
+	FltBhvType NATBhvType // FilteringBehaviorType
+}
```
Comment (AudriusButkevicius) [inline stun/const.go:38]: Whats up with these cryptic names? Are you paying per character?

Comment (kevinlemonra) [reply stun/const.go:38]: Thanks for response.   It's just some ridiculous thoughts: don't want it too long. Is it better use MType and FType?

Comment (kevinlemonra) [reply stun/const.go:38]: Should I move these to a test branch?  There may need much more adjust.

**Comment by kevinlemonra** [comment]
Just want add the behavior test mode, so remove other stuff.

File: stun/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevinlemonra) [reply stun/client.go]: Your are right,  no need to change it.

**Comment by kevinlemonra** [comment]
Here is some test.

2022/01/22 13:52:33 Do Test1
2022/01/22 13:52:33 Send To: 106.55.201.202:3478
2022/01/22 13:52:34 Received: {packet nil: false, local: 112.97.57.166:49144, remote: 106.55.201.202:3478, changed: 106.55.201.64:8000, other: <nil>, identical: false}
2022/01/22 13:52:34 Do Test2
2022/01/22 13:52:34 Send To: 106.55.201.64:3478
2022/01/22 13:52:34 Received: {packet nil: false, local: 112.97.57.166:49144, remote: 106.55.201.64:3478, changed: 106.55.201.202:8000, other: <nil>, identical: false}
2022/01/22 13:52:34 Do Test4
2022/01/22 13:52:34 Send To: 106.55.201.64:8000
2022/01/22 13:52:34 Received: {packet nil: false, local: 112.97.57.166:49144, remote: 106.55.201.202:3478, changed: 106.55.201.202:3478, other: <nil>, identical: false}
  Mapping Behavior: EndpointIndependent
Filtering Behavior: EndpointIndependent
   Normal NAT Type: Full cone NAT


2022/01/22 13:39:58 Do Test1
2022/01/22 13:39:58 Send To: 18.191.223.12:3478
2022/01/22 13:39:58 Received: {packet nil: false, local: 112.97.57.166:38217, remote: 18.191.223.12:3478, changed: <nil>, other: 18.219.110.18:3479, identical: false}
2022/01/22 13:39:58 Do Test2
2022/01/22 13:39:58 Send To: 18.219.110.18:3478
2022/01/22 13:39:58 Received: {packet nil: false, local: 112.97.61.166:37519, remote: 18.219.110.18:3478, changed: <nil>, other: 18.191.223.12:3479, identical: false}
2022/01/22 13:39:58 Do Test3
2022/01/22 13:39:58 Send To: 18.219.110.18:3479
2022/01/22 13:39:59 Received: {packet nil: false, local: 112.97.61.166:37519, remote: 18.219.110.18:3479, changed: <nil>, other: 18.191.223.12:3478, identical: false}
2022/01/22 13:39:59 Do Test4
2022/01/22 13:39:59 Send To: 18.191.223.12:3478
2022/01/22 13:39:59 Received: {packet nil: false, local: 112.97.57.166:38217, remote: 18.219.110.18:3479, changed: <nil>, other: 18.219.110.18:3479, identical: false}
  Mapping Behavior: AddressDependent
Filtering Behavior: EndpointIndependent
   Normal NAT Type: Undefined


2022/01/22 15:02:50 Do Test1
2022/01/22 15:02:50 Send To: 106.55.201.64:3478
2022/01/22 15:02:50 Received: {packet nil: false, local: 113.87.88.208:55899, remote: 106.55.201.64:3478, changed: 106.55.201.202:8000, other: <nil>, identical: false}
2022/01/22 15:02:50 Do Test2
2022/01/22 15:02:50 Send To: 106.55.201.202:3478
2022/01/22 15:02:50 Received: {packet nil: false, local: 113.87.88.208:55899, remote: 106.55.201.202:3478, changed: 106.55.201.64:8000, other: <nil>, identical: false}
2022/01/22 15:02:50 Do Test4
2022/01/22 15:02:50 Send To: 106.55.201.64:3478
2022/01/22 15:02:59 Received: Nil
2022/01/22 15:02:59 Do Test5
2022/01/22 15:02:59 Send To: 106.55.201.64:3478
2022/01/22 15:03:09 Received: Nil
  Mapping Behavior: EndpointIndependent
Filtering Behavior: AddressAndPortDependent
   Normal NAT Type: Port Restricted cone NAT


2022/01/22 07:20:37 Send To: 111.206.174.2:3478
2022/01/22 07:20:37 Received: {packet nil: false, local: 113.87.88.208:39607, remote: 111.206.174.2:3478, changed: <nil>, other: 111.206.174.3:3479, identical: true}
Not behind a NAT.

File: stun/const.go (hunk 2 above)
Comment (AudriusButkevicius) [reply stun/const.go:37]: The type is already called NATBehaviour, why can't you just call them FilteringType and BehaviourType?

Comment (kevinlemonra) [reply stun/const.go:37]: Good advice.

**Review by kevinlemonra — COMMENTED** [review]
Change to MappingType and FilteringType.

File: stun/discover.go [hunk 3]
```diff
@@ -164,3 +164,93 @@ func (c *Client) discover(conn net.PacketConn, addr *net.UDPAddr) (NATType, *Hos
...
+	// Test2   ->(IP2,port1)
+	// Perform test to see if mapping to the same IP and port when
+	// send to another IP.
+	c.logger.Debugln("Do Test2")
+	tmpAddr := &net.UDPAddr{IP: net.ParseIP(otherAddr.IP()), Port: addr.Port}
+	resp2, err := c.test(conn, tmpAddr)
+	if err != nil {
+		return nil, err
+	}
+	if resp2.mappedAddr.IP() == resp1.mappedAddr.IP() &&
```
Comment (ccding) [inline stun/discover.go:204]: got the following error here
```
➜  go-stun git:(test) ./go-stun -b
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x110ead2]

goroutine 1 [running]:
github.com/ccding/go-stun/stun.(*Client).behaviorTest(0xc000159ef8, 0x117b480, 0xc00012a028, 0xc000114780, 0x0, 0x0, 0x0)
        /Users/cong/go/src/github.com/ccding/go-stun/stun/discover.go:204 +0x232
github.com/ccding/go-stun/stun.(*Client).BehaviorTest(0xc00003eef8, 0x0, 0x0, 0x0)
        /Users/cong/go/src/github.com/ccding/go-stun/stun/client.go:120 +0xe5
main.behaviorTest(0xc00003eef8)
        /Users/cong/go/src/github.com/ccding/go-stun/main.go:61 +0x45
main.main()
        /Users/cong/go/src/github.com/ccding/go-stun/main.go:41 +0x6a5
```

Comment (ccding) [reply stun/discover.go:204]: Also `local: 62.24.197.181:1298` is not my local IP address

Comment (kevinlemonra) [reply stun/discover.go:204]: It should be nat blocked , but will be detected in the first step.Maybe the network is not stable.

Comment (kevinlemonra) [reply stun/discover.go:201]: > Also `local: 62.24.197.181:1298` is not my local IP address

Is it better to use  mapped   instead of  local?

**Review by ccding — APPROVED** [review]

**Pull request merged**